azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -526,3 +526,184 @@ def _handle_pipeline_event(self, event):
+        """
+        Return True if this stage needs to put a timer on this operation
+        """
+        return op.__class__ in self.timeout_intervals
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        if self._should_watch_for_timeout(op):
+            # Create a timer to watch for operation timeout on this op and attach it
+            # to the op.
+            self_weakref = weakref.ref(self)
+
+            @pipeline_thread.invoke_on_pipeline_thread_nowait
+            def on_timeout():
+                this = self_weakref()
+                logger.info("{}({}): returning timeout error".format(this.name, op.name))
+                this._complete_op(
+                    op,
+                    transport_exceptions.PipelineTimeoutError(
+                        "operation timed out before protocol client could respond"
+                    ),
+                )
+
+            logger.debug("{}({}): Creating timer".format(self.name, op.name))
+            op.timeout_timer = Timer(self.timeout_intervals[op.__class__], on_timeout)
+            op.timeout_timer.start()
+
+            # Send the op down, but intercept the return of the op so we can
+            # remove the timer when the op is done
+            logger.debug("{}({}): Sending down".format(self.name, op.name))
+            self._send_op_down_and_intercept_return(
+                op=op, intercepted_return=self._on_intercepted_return
+            )
+        else:
+            self._send_op_down(op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _on_intercepted_return(self, op, error):
+        # When an op comes back, delete the timer and pass it right up.
+        del op.timeout_timer
+        logger.debug("{}({}): Op completed".format(self.name, op.name))
+        self._send_completed_op_up(op, error)
+
+
+class RetryStage(PipelineStage):
+    """
+    The purpose of the retry stage is to watch specific operations for specific
+    errors and retry the operations as appropriate.
+
+    This stage is currently very specific as it watches a hardcoded set of
+    operations for a hardcoded set of errors and retries after a hardcoded
+    period of time.
+
+    Like the TimeoutStage, the hardcoded nature of this stage will eventually
+    be provided by initialization parameters, and there will likely be two
+    retry stages which can handle IoTHub and MQTT retries at different levels
+    of the pipeline.
+
+    Unlike the TimeoutStage, this stage will never need to worry about cancelling
+    failed operations.  When an operation is retried at this stage, it is already
+    considered "failed", so no cancellation needs to be done.
+    """
+
+    def __init__(self):
+        super(RetryStage, self).__init__()
+        self.retry_intervals = {
+            pipeline_ops_mqtt.MQTTSubscribeOperation: 20,
+            pipeline_ops_mqtt.MQTTUnsubscribeOperation: 20,
+        }
+        self.waiting_to_retry = []
```

Comment:
>waiting_to_retry [](start = 13, length = 16)

Is this a list of ops? If so, could we call this `ops_waiting_to_retry` for clarity?